plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/query/HudiReadOptimizedDirectoryLister.java

```diff
@@ -62,63 +54,43 @@ public HudiReadOptimizedDirectoryLister(
             HoodieTableMetaClient metaClient,
             HiveMetastore hiveMetastore,
             Table hiveTable,
-            List<HiveColumnHandle> partitionColumnHandles)
+            List<HiveColumnHandle> partitionColumnHandles,
+            List<String> hivePartitionNames)
     {
-        this.tableHandle = tableHandle;
-        this.tableName = tableHandle.getSchemaTableName();
-        this.hiveMetastore = hiveMetastore;
-        this.hiveTable = hiveTable;
-        this.partitionColumnHandles = partitionColumnHandles;
         this.fileSystemView = FileSystemViewManager.createInMemoryFileSystemView(engineContext, metaClient, metadataConfig);
-        this.partitionKeysFilter = MetastoreUtil.computePartitionKeyFilter(partitionColumnHandles, tableHandle.getPartitionPredicates());
+        this.latestInstant = metaClient.getActiveTimeline().getCommitsTimeline().filterCompletedInstants().lastInstant().map(HoodieInstant::getTimestamp).orElse(null);
         this.partitionColumns = hiveTable.getPartitionColumns();
-    }
-
-    @Override
-    public List<HudiPartitionInfo> getPartitionsToScan()
-    {
-        if (hivePartitionNames == null) {
-            hivePartitionNames = partitionColumns.isEmpty()
-                    ? Collections.singletonList("")
-                    : getPartitionNamesFromHiveMetastore(partitionKeysFilter);
-        }
-
-        List<HudiPartitionInfo> allPartitionInfoList = hivePartitionNames.stream()
-                .map(hivePartitionName -> new HiveHudiPartitionInfo(
-                        hivePartitionName,
-                        partitionColumns,
-                        partitionColumnHandles,
-                        tableHandle.getPartitionPredicates(),
-                        hiveTable,
-                        hiveMetastore))
-                .collect(Collectors.toList());
-
-        return allPartitionInfoList.stream()
-                .filter(partitionInfo -> partitionInfo.getHivePartitionKeys().isEmpty() || partitionInfo.doesMatchPredicates())
-                .collect(Collectors.toList());
+        this.allPartitionInfoMap = hivePartitionNames.stream()
+                .collect(Collectors.toMap(
+                        Function.identity(),
+                        hivePartitionName -> new HiveHudiPartitionInfo(
+                                hivePartitionName,
+                                partitionColumns,
+                                partitionColumnHandles,
+                                tableHandle.getPartitionPredicates(),
+                                hiveTable,
+                                hiveMetastore)));
     }
 
     @Override
     public List<FileStatus> listStatus(HudiPartitionInfo partitionInfo)
     {
+        if (latestInstant != null) {
+            return fileSystemView.getLatestFileSlicesBeforeOrOn(partitionInfo.getRelativePartitionPath(), latestInstant, false)
```

Comment:
This seems like new functionality, unrelated to the concurrency change.

Since we're expanding the usage of `FileSystemViewManager`, could you add another commit to this PR as a precursor that switches `HudiReadOptimizedDirectoryLister` to use `TrinoFileSystem`?

Reply:
`TrinoFileSystem` can be used list a partition but a partition in Hudi could contain older file versions as well and to filter those out we need `FileSystemViewManager`. It is more efficient to first filter latest version (based on Hudi commit metadata in [timeline](https://hudi.apache.org/docs/timeline)) and then list the remaining files. 
That said, we can still remove the usage of `FileStatus` from `HudiDrirectoryLister`. I'll create a separate PR for that.

Reply:
I've rebased with latest master and used `HudiTableFileSystemView` instead.